### PR TITLE
[subset] Hmtx/vmtx tables to use subset2 and new iterator frameworks

### PR DESCRIPTION
**Pull request by qxliu76**

### CODE REVIEW

File: src/hb-ot-hmtx-table.hh [hunk 1]
```diff
@@ -86,74 +92,68 @@ struct hmtxvmtx
...
+    _mtx.init (c->plan->source);
+    unsigned num_advances = _mtx.num_advances_for_subset (c->plan);
+    
+    auto it = 
+    + hb_range (c->plan->num_output_glyphs ())
+    | hb_map ([c, &_mtx] (unsigned _)
+	{
+	  hb_codepoint_t old_gid;
+	  if (c->plan->old_gid_for_new_gid (_, &old_gid))
+            return hb_pair<unsigned, unsigned> (_mtx.get_advance (old_gid), _mtx.get_side_bearing (old_gid));
```
Comment (behdad) [inline src/hb-ot-hmtx-table.hh]: Do you actually need to specify the template types here?  I'd assume not.

File: src/hb-ot-hmtx-table.hh [hunk 2]
```diff
@@ -86,74 +92,68 @@ struct hmtxvmtx
...
+          else
+	    return hb_pair<unsigned, unsigned> (0, 0);
```
Comment (behdad) [inline src/hb-ot-hmtx-table.hh]: Same here.  Just pass `0u` instead of `0`.

File: src/hb-subset.cc [hunk 3]
```diff
@@ -67,7 +67,7 @@ template<typename TableType>
 static bool
 _subset2 (hb_subset_plan_t *plan)
 {
-  bool result = true;
+  bool result = false;
```
Comment (behdad) [inline src/hb-subset.cc]: Why this?

File: src/hb-ot-hmtx-table.hh [hunk 4]
```diff
@@ -51,6 +51,12 @@ struct LongMetric
   FWORD		sb; /* Leading (left/top) side bearing. */
   public:
   DEFINE_SIZE_STATIC (4);
+  LongMetric (unsigned ad,
+              unsigned side_bearing)
+  {
+    advance = ad;
+    sb = side_bearing;
+  }
```
Comment (behdad) [inline src/hb-ot-hmtx-table.hh]: We actually don't currently have constructors in any of our font structs.  Previously that was a requirement of C++<11.  I still want to keep it that way.  So just assign to the members outside for now.  Thanks.

File: src/hb-ot-hmtx-table.hh [hunk 5]
```diff
@@ -86,74 +92,68 @@ struct hmtxvmtx
...
+    unsigned idx = 0;
+    + it
+    | hb_apply ([c, &idx, num_advances] (const hb_item_type<Iterator>& _)
+                {
+                  if (idx < num_advances) 
+                  {
+                    LongMetric lm (_.first, _.second);
+                    if (unlikely (!c->embed<LongMetric> (&lm)))
+                      return;
+                  } else {
```
Comment (behdad) [inline src/hb-ot-hmtx-table.hh]: Braces on new line.

File: src/hb-ot-hmtx-table.hh [hunk 6]
```diff
@@ -86,74 +92,68 @@ struct hmtxvmtx
...
-    char * dest_pos = (char *) dest;
+    unsigned idx = 0;
+    + it
+    | hb_apply ([c, &idx, num_advances] (const hb_item_type<Iterator>& _)
+                {
+                  if (idx < num_advances) 
+                  {
+                    LongMetric lm (_.first, _.second);
+                    if (unlikely (!c->embed<LongMetric> (&lm)))
+                      return;
```
Comment (behdad) [inline src/hb-ot-hmtx-table.hh]: Return on same line.

File: src/hb-subset.cc (hunk 3 above)
Comment (qxliu76) [reply src/hb-subset.cc]: this is to be consistent with _subset(), otherwise there's a unit test failure.

**Pull request merged**